### PR TITLE
Reconnect - "re-establish a bond of communication or emotion"

### PR DESCRIPTION
**Pull request by moozzyk**
Adding support for reconnecting to the server if the connection is lost.

### CODE REVIEW

**Comment by moozzyk** [comment]
@halter73 - some food for your multi-threaded brain (if you would be so kind)

File: src/signalrclient/connection_impl.cpp [hunk 1]
```diff
@@ -397,13 +442,188 @@ namespace signalr
...
+                        connection->m_logger.log(
+                            trace_level::errors,
+                            utility::string_t(_XPLATSTR("reconnected callback threw an exception: "))
+                            .append(utility::conversions::to_string_t(e.what())));
+                    }
+                    catch (...)
+                    {
+                        connection->m_logger.log(
+                            trace_level::errors,
+                            utility::string_t(_XPLATSTR("reconnected callback threw an unknown exception")));
```
Comment (BrennanConroy) [inline src/signalrclient/connection_impl.cpp]: nit: You removed the explicit constructor call to utility::string_t earlier in this PR but have it here

Comment (moozzyk) [inline src/signalrclient/connection_impl.cpp]: will fix

File: test/signalrclienttests/connection_impl_tests.cpp [hunk 2]
```diff
@@ -875,6 +917,70 @@ TEST(connection_impl_stop, stop_ignores_exceptions_from_abort_requests)
...
+    connection->set_disconnected([&disconnected_invoked](){ disconnected_invoked = true; });
+
+    connection->start()
+        .then([connection]()
+    {
+        return connection->stop();
+    })
+        .then([connection]()
+    {
+        return connection->start();
```
Comment (BrennanConroy) [inline test/signalrclienttests/connection_impl_tests.cpp]: This second start isn't needed

**Comment by BrennanConroy** [comment]
:shipit:

File: src/signalrclient/connection_impl.cpp [hunk 3]
```diff
@@ -345,6 +368,24 @@ namespace signalr
                     // we do let the exception through (especially the task_canceled exception)
                     connection->m_transport = nullptr;
                 }
+
+                try
+                {
+                    connection->m_disconnected();
```
Comment (moozzyk) [inline src/signalrclient/connection_impl.cpp]: This is now called when we hold the lock - we should release the lock before calling user's code.

**Comment by halter73** [comment]
:shipit:

File: test/signalrclienttests/connection_impl_tests.cpp (hunk 2 above)
Comment (moozzyk) [inline test/signalrclienttests/connection_impl_tests.cpp]: Fixed

**Comment by moozzyk** [comment]
Thanks @BrennanConroy and @halter73 !

**Pull request merged**